Ports prefixScan, OneToManyAssoc and HistoContainer from CUDAUtilities.

ericcano opened the pull request:
#### PR description:

This PR brings and ports the latest developments from @VinInn on PrefixScan, OneToManyAssoc and HistoContainer from `HeterogeneousCore/CUDAUtilities/` to Alpaka.
Latest features are:
- Single kernel prefix scan
- Optional runtime sized ones array, many array and content array for the HistoContainer
- Reworked dependency order, with the HistoContainer built on top of the OneToManyAssoc.

#### PR validation:

Unit tests were also ported and pass

ericcano committed:
Ports prefixScan, OneToManyAssoc and HistoContainer from CUDAUtilities.

OneToManyAssoc is also separated into various function depending on use case.

Co-authored-by: Andrea Bocci <fwyzard@gmail.com>

ericcano commented:
@VinInn, @AdrianoDee, FYI
I will now create the class hierarchy based on this.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37266

- This PR adds an extra 36KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41284, #41564
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41284

cmsbuild commented:
A new Pull Request was created by  @ericcano (Eric Cano) for master.

It involves the following packages:

- HeterogeneousCore/AlpakaInterface (**heterogeneous**)


@cmsbuild, @fwyzard, @makortel can you please review it and eventually sign? Thanks.
@missirol, @makortel, @rovere this is something you requested to watch as well.
@antoniovilela, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37270

- This PR adds an extra 36KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117, #41284, #41564
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117, #41284

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37270/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37270/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

makortel commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,233 @@
+#include <algorithm>
+#include <cstdint>
+
+#include <time.h>
```
Why is this `#include` needed?

ericcano replied:
No. Debugging leftover, I cleared them.

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc:
```diff
@@ -0,0 +1,110 @@
+TEST_CASE("Standard checks of " ALPAKA_TYPE_ALIAS_NAME(alpakaTestAtomicPair), s_tag) {
+  SECTION("AtomicPairCounter") {
+    auto const &devices = cms::alpakatools::devices<Platform>();
+    if (devices.empty()) {
+      std::cout << "No devices available on the platform " << EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE)
+                << ", the test will be skipped.\n";
+      return;
+    }
```
Given the somewhat recent changes to how scram runs the unit tests for the alpaka backends (i.e. e.g. cuda tests are ran only if `cudaIsEnabled` program returned `0`), I'd argue this way of skipping the test in absence of device is not needed, and it would actually be better if the test fails in absence of the device(s).

ericcano replied:
Without this section, we still get a success (but with no message) as the tests run inside this loop:

```C++
for (auto const &device : devices) {
```
Another option would be to explicitly fail (```assert(!devices.empty())``` ?). Is this a more standard behavior?

makortel replied:
I think it would be good for these tests to require a device (but I'd use catch2's `REQUIRE()` rather than `assert()`).

fwyzard replied:
> Given the somewhat recent changes to how scram runs the unit tests for the alpaka backends (i.e. e.g. cuda tests are ran only if `cudaIsEnabled` program returned `0`), I'd argue this way of skipping the test in absence of device is not needed, and it would actually be better if the test fails in absence of the device(s).

How do these changes work?
Should we update all existing tests accordingly?

makortel replied:
> > Given the somewhat recent changes to how scram runs the unit tests for the alpaka backends (i.e. e.g. cuda tests are ran only if `cudaIsEnabled` program returned `0`), I'd argue this way of skipping the test in absence of device is not needed, and it would actually be better if the test fails in absence of the device(s).
> 
> How do these changes work?

The behavior is described in https://github.com/cms-sw/cmssw/issues/41340#issuecomment-1514867660

> Should we update all existing tests accordingly?

I already did privately for `HeterogeneousCore/AlpakaTest` last week (just need to rebase now that the preceding PRs got merged).

About direct CUDA I'm not sure. Ideally yes, but is it worth of the effort?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc:
```diff
@@ -0,0 +1,212 @@
+    auto queue = Queue(device);
+    const auto warpSize = alpaka::getWarpSizes(device)[0];
+    // WARP PREFIXSCAN (OBVIOUSLY GPU-OqNLY)
+#if defined(ALPAKA_ACC_GPU_CUDA_ENABLED) || defined(ALPAKA_ACC_GPU_HIP_ENABLED)
```
Maybe consider `if constexpr`?

ericcano replied:
Yes

fwyzard replied:
why?
`ALPAKA_ACC_GPU_CUDA_ENABLED` _etc_ are macros, isn't it more consistent to use `#if` with them ?

fwyzard replied:
ah, after reading the comments above, do you mean to use the trait that checks for a single thread ?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc:
```diff
@@ -0,0 +1,255 @@
+#ifndef NDEBUG
+      auto bk = h->bin(v[k]);
+#endif
+      ALPAKA_ASSERT_OFFLOAD(bk == i);
```
It is not clear to me if in the current setup `ALPAKA_ASSERT_OFFLOAD` actually expands to an assertion (or similar) on all backends. I see @fwyzard wondered this out loud in https://github.com/alpaka-group/alpaka/issues/2001 . In addition, in the `CUDAUtilities` all the tests are compiled with `-DGPU_DEBUG` to make sure the assertions are done also on GPU (in conjunction with
https://github.com/cms-sw/cmssw/blob/e4bfeac8c9a8d7e1690b8a16cd8464e0330b24dd/HeterogeneousCore/CUDAUtilities/interface/cuda_assert.h#L4-L16
)

fwyzard replied:
Agreed... maybe we can converge on what the _desired_ behaviour for `ALPAKA_ASSERT_OFFLOAD` should be, and then we can propose it to the Alpaka developers, or implement our own variant.

fwyzard replied:
See https://github.com/alpaka-group/alpaka/issues/2186 .

makortel reviewed (commented):
A cursory look

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h:
```diff
@@ -0,0 +1,203 @@
+    };
+
+    template <typename TAcc, typename Histo, typename T, typename TQueue>
+    inline __attribute__((always_inline)) void fillManyFromVector(Histo *__restrict__ h,
```
Yes, it should, will look after those missed alpaka-isations.

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,273 @@
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h"
+
+namespace {
```
This is a workaround for signed/unsigned indices comparisons. I'll try to find a better solution.

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,273 @@
+        ALPAKA_ASSERT_OFFLOAD(view.offStorage);
+        ALPAKA_ASSERT_OFFLOAD(view.offSize > 0);
+      }
+#if !defined(ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED)
+      auto nthreads = 1024;
+      auto nblocks = 1;  // MUST BE ONE as memory is initialize in thread 0 (alternative is two kernels);
+      auto workDiv = cms::alpakatools::make_workdiv<TAcc>(nblocks, nthreads);
+      alpaka::exec<TAcc>(queue, workDiv, zeroAndInit{}, view);
+#else
+      auto h = view.assoc;
+      ALPAKA_ASSERT_OFFLOAD(h);
+      h->initStorage(view);
+      h->zero();
+      h->psws = 0;
+#endif
```
I used the latter in places, this should be generalized indeed.

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,273 @@
+    inline __attribute__((always_inline)) void launchFinalize(OneToManyAssocView<Assoc> view, TQueue &queue) {
+      auto h = view.assoc;
+      ALPAKA_ASSERT_OFFLOAD(h);
+#if !defined(ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED)
```
Yes

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,273 @@
+      constexpr index_type const *end(uint32_t b) const { return content.data() + off[b + 1]; }
+
+      FlexiStorage<Counter, ONES> off;
+      int32_t psws;  // prefix-scan working space
+      FlexiStorage<index_type, SIZE> content;
```
Yep, can do.

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h:
```diff
@@ -0,0 +1,35 @@
+#include <alpaka/alpaka.hpp>
+
+// reimplementation of std algorithms able to compile with Alpaka,
+// mostly by declaring them constexpr
```
Yes

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,233 @@
+          ws = alpaka::getDynSharedMem<T>(acc);
+        }
+        ALPAKA_ASSERT_OFFLOAD(warpSize == static_cast<std::size_t>(alpaka::warp::getSize(acc)));
+#ifdef __CUDA_ARCH__
+        // TODO ALPAKA_ASSERT_OFFLOAD(sizeof(T) * gridDim.x <= dynamic_smem_size());  // size of psum below
+#endif
```
No, and it's CUDA specific.

fwyzard replied:
should we look for a way to implement the corresponding check with alpaka ?

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,233 @@
+        [[maybe_unused]] const auto threadsPerBlock = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
+        [[maybe_unused]] const auto blocksPerGrid = alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+        [[maybe_unused]] const auto blockIdx = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+        [[maybe_unused]] const auto threadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];
```
This was for debugging. I removed really unused variables, untaged the other ones,  except one which is a needed shortcut (the `ALPAKA_ASSERT_OFFLOAD` macro it is used into chokes on the comma if we use the value directly, and the variable is indeed unused when `ALPAKA_ASSERT_OFFLOAD` expends to nothing).

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37362

- This PR adds an extra 40KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117, #41284

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37362/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37362/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37364

- This PR adds an extra 76KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117, #41284

cmsbuild commented:
Pull request #43064 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,223 @@
+        ALPAKA_ASSERT_OFFLOAD(size <= alpaka::warp::getSize(acc) * warpSize);
+        ALPAKA_ASSERT_OFFLOAD(0 == blockDimension % warpSize);
+        auto first = blockThreadIdx;
+        auto laneId = blockThreadIdx & (warpSize - 1);
```
This requires that `warpSize` be a power of 2.
So far, all implementations I know of meet this requirement.
Would it make sense to check that ?

fwyzard replied:
Or change it to
```suggestion
        auto laneId = blockThreadIdx % warpSize;
```
?

ericcano replied:
Cannot be static, of course, so made it a runtime (but debug version only) assertion.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,223 @@
+        for (auto i = first; i < size; i += blockDimension) {
+          warpPrefixScan(acc, laneId, ci, co, i);
+          auto warpId = i / warpSize;
+          // FIXME test ?
```
what is there to fix ?

ericcano replied:
Looks fine, I will remove the comment.

ericcano commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,223 @@
+        int32_t const blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        int32_t const blockThreadIdx(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        ALPAKA_ASSERT_OFFLOAD(ws);
+        ALPAKA_ASSERT_OFFLOAD(size <= alpaka::warp::getSize(acc) * warpSize);
```
Indeed!

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,223 @@
+  namespace alpakatools {
+
+    template <typename TAcc, typename T, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE void warpPrefixScan(const TAcc& acc, int32_t laneId, T const* ci, T* co, uint32_t i) {
```
The CUDA version computes the lane id internally; was there a particular reason to pass it as a parameter instead ?

ericcano replied:
The warp size goes from a compile time value (hardcoded value 32) to a computed value based on  a variable, so I minimized recomputations.

fwyzard replied:
ah, I understand now, thanks !

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37387

- This PR adds an extra 76KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117, #41284
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117, #41284

cmsbuild commented:
Pull request #43064 was updated. @makortel, @fwyzard, @cmsbuild can you please check and sign again.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+                                    uint32_t size,
+                                    int32_t numBlocks,
+                                    int32_t* pc,
+                                    std::size_t warpSize) const {
```
`warpSize` is passed as an argument because it's needed by `BlockSharedMemDynSizeBytes::getBlockSharedMemDynSizeBytes(...)` ?

ericcano replied:
Yes, and also used in the kernel itself, where we could get it through the accelerator instance, but not in the memory sizing trait.

fwyzard replied:
Ah, that's a very good point.

Could you add a comment to remind about this detail ?
Otherwise I'm worried we will forget why we need it as an argument here, and spend time trying to remove it.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+                                    int32_t numBlocks,
+                                    int32_t* pc,
+                                    std::size_t warpSize) const {
+        // Get shared variable. TODO: The workspace is in fact not needed at all for single thread multi element accelerators TODO
```
I think this is properly implements, and so `TODO` can be dropped from the comment ?
```suggestion
        // Get shared variable. The workspace is needed only for multi-threaded accelerators.
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+
+        // count blocks that finished
+        auto& isLastBlockDone = alpaka::declareSharedVar<bool, __COUNTER__>(acc);
+        //__shared__ bool isLastBlockDone;
```
I think this comment can now be removed ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_prefixScan_h
+#define HeterogeneousCore_AlpakaInterface_interface_prefixScan_h
+// Active version
```
Maybe this should be removed ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
```
```suggestion
#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"

```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
+namespace cms {
+  namespace alpakatools {
```
```suggestion
namespace cms::alpakatools {
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+namespace cms {
+  namespace alpakatools {
+
+    template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+    constexpr bool isPowerOf2(T v) {
+      // returns true iif v has only one bit set.
+      while (v) {
+        if (v & 1)
+          return !(v >> 1);
+        else
+          v >>= 1;
+      }
+      return false;
+    }
```
This might be faster, since it is not iterative:
```c++
    // returns true if v is positive and is a power of two
    template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
    constexpr bool isPowerOf2(T v) {
      return (v > 0) and (v & v-1) == 0;
    }
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+        const TAcc& acc, T const* ci, T* co, int32_t size, T* ws = nullptr) {
+      if constexpr (!requires_single_thread_per_block_v<TAcc>) {
+        const auto warpSize = alpaka::warp::getSize(acc);
+        int32_t const blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        int32_t const blockThreadIdx(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
```
```suggestion
        const int32_t blockDimension = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
        const int32_t blockThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];
```
for consistency with the previous line.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+        ALPAKA_ASSERT_OFFLOAD(ws);
+        ALPAKA_ASSERT_OFFLOAD(size <= warpSize * warpSize);
+        ALPAKA_ASSERT_OFFLOAD(0 == blockDimension % warpSize);
+        auto first = blockThreadIdx;
```
`const`

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+        ALPAKA_ASSERT_OFFLOAD(0 == blockDimension % warpSize);
+        auto first = blockThreadIdx;
+        ALPAKA_ASSERT_OFFLOAD(isPowerOf2(warpSize));
+        auto laneId = blockThreadIdx & (warpSize - 1);
```
```suggestion
        const auto laneId = blockThreadIdx & (warpSize - 1);
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+                                                             T* __restrict__ ws = nullptr) {
+      if constexpr (!requires_single_thread_per_block_v<TAcc>) {
+        const auto warpSize = alpaka::warp::getSize(acc);
+        int32_t const blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        int32_t const blockThreadIdx(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
```
const int32_t blockDimension = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
        const int32_t blockThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];

fwyzard replied:
```suggestion
        const int32_t blockDimension = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
        const int32_t blockThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+        ALPAKA_ASSERT_OFFLOAD(ws);
+        ALPAKA_ASSERT_OFFLOAD(size <= warpSize * warpSize);
+        ALPAKA_ASSERT_OFFLOAD(0 == blockDimension % warpSize);
+        auto first = blockThreadIdx;
+        auto laneId = blockThreadIdx & (warpSize - 1);
```
```suggestion
        const auto first = blockThreadIdx;
        ALPAKA_ASSERT_OFFLOAD(isPowerOf2(warpSize));
        const auto laneId = blockThreadIdx & (warpSize - 1);
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+        const auto threadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];
+        ALPAKA_ASSERT_OFFLOAD(elementsPerGrid >= size);
+        // first each block does a scan
+        [[maybe_unused]] int off = elementsPerBlock * blockIdx;
```
why `[[maybe_unused]]` ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+        auto& isLastBlockDone = alpaka::declareSharedVar<bool, __COUNTER__>(acc);
+        //__shared__ bool isLastBlockDone;
+        if (0 == threadIdx) {
+          alpaka::mem_fence(acc, alpaka::memory_scope::Device{});
```
```suggestion
          alpaka::mem_fence(acc, alpaka::memory_scope::Grid{});
```

`Grid` should be enough, we don't need to have a fence visible to other kernel launches, right ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/prefixScan.h:
```diff
@@ -0,0 +1,235 @@
+      }
+    };
+
+  }  // namespace alpakatools
+}  // namespace cms
```
```suggestion
}  // namespace cms::alpakatools
```

fwyzard commented:
I should be done with `HeterogeneousCore/AlpakaInterface/interface/prefixScan.h`.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h:
Other than the namespace, this class is identical to `cms::cuda::FlexiStorage` - and neither depends on CUDA or Alpaka.
@makortel, maybe we could move them to a more general place, like `DataFormats/Common` or `FWCore/Utilities`, and into the `edm` namespace ?

makortel replied:
The generic `FlexiStorage` looks like an `std::array`, but with smaller interface. The specialization for `-1` capacity looks like a way to deal with runtime-defined capacity. Together, for now at least, they look like an implementation detail of `OneToManyAssocBase`. To me this looks (currently) too specific use case to be moved to `DataFormats/Common` or `FWCore/Utilities`. The (duplicated) CUDA version should anyway disappear in the "near future".

fwyzard commented:
@makortel @ericcano - _not for this PR_ but later on - maybe we should review `SimpleVector`, `VecArray` and `FlexiStorage` ?
They all seem somewhat incomplete and overlapping among themselves in terms of features and interface.

fwyzard commented:
Instead, @ericcano, cold I ask you to try to write some documentation for the tools you have ported from CUDA to Alpaka and included in this PR ?

I _know_ we should have done it earlier already for the CUDA case, but I don't think we did, and you are my best hope for this at the moment :-)

makortel commented:
> @makortel @ericcano - _not for this PR_ but later on - maybe we should review `SimpleVector`, `VecArray` and `FlexiStorage` ? They all seem somewhat incomplete and overlapping among themselves in terms of features and interface.

I agree the current situation is confusing, especially for anyone new to figure out what to use.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h:
```diff
@@ -0,0 +1,203 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_HistoContainer_h
+#define HeterogeneousCore_AlpakaInterface_interface_HistoContainer_h
+
+#include <alpaka/alpaka.hpp>
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/prefixScan.h"
+
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
```
```suggestion
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <type_traits>

#include <alpaka/alpaka.hpp>

#include "HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h"
#include "HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h"
#include "HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h"
#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
#include "HeterogeneousCore/AlpakaInterface/interface/prefixScan.h"
#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"

```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h:
```diff
@@ -0,0 +1,203 @@
+
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+namespace cms {
+  namespace alpakatools {
```
```suggestion
namespace cms::alpakatools {
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h:
```diff
@@ -0,0 +1,203 @@
+                                    T const *__restrict__ v,
+                                    uint32_t const *__restrict__ offsets) const {
+        const uint32_t nt = offsets[nh];
+        for_each_element_in_grid_strided(acc, nt, [&](uint32_t i) {
```
does it make any difference in performance if you use
```suggestion
        for (uint32_t i: elements_with_stride(acc, nt)) {
```
?

ericcano replied:
Most probably not, with a slight chance for the range version to be better. I moved to the range version so the PR does not need to introduce the for_each anymore.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h:
```diff
@@ -0,0 +1,203 @@
+    template <typename TAcc, typename Histo, typename T, typename TQueue>
+    ALPAKA_FN_INLINE void fillManyFromVector(Histo *__restrict__ h,
+                                             uint32_t nh,
+                                             T const *v,
+                                             uint32_t const *offsets,
```
If I'm not mistaken, in the CUDA version these were `__restrict__`.
Is it different on purpose here ?

ericcano replied:
No, it's a miss, I put them back.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h:
```diff
@@ -0,0 +1,203 @@
+      const auto blocksPerGrid = divide_up_by(totSize, nthreads);
+      const auto workDiv = make_workdiv<TAcc>(blocksPerGrid, threadsPerBlockOrElementsPerThread);
+
+      alpaka::exec<TAcc>(queue, workDiv, /* Histo:: */ countFromVector(), h, nh, v, offsets);
```
```suggestion
      alpaka::exec<TAcc>(queue, workDiv, countFromVector(), h, nh, v, offsets);
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h:
```diff
@@ -0,0 +1,203 @@
+      alpaka::exec<TAcc>(queue, workDiv, /* Histo:: */ countFromVector(), h, nh, v, offsets);
+      Histo::template launchFinalize<TAcc>(h, queue);
+
+      alpaka::exec<TAcc>(queue, workDiv, /* Histo:: */ fillFromVector(), h, nh, v, offsets);
```
```suggestion
      alpaka::exec<TAcc>(queue, workDiv, fillFromVector(), h, nh, v, offsets);
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,276 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_OneToManyAssoc_h
+#define HeterogeneousCore_AlpakaInterface_interface_OneToManyAssoc_h
+
+#include <alpaka/alpaka.hpp>
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/prefixScan.h"
+
+#include "HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h"
```
```suggestion
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <type_traits>

#include <alpaka/alpaka.hpp>

#include "HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h"
#include "HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h"
#include "HeterogeneousCore/AlpakaInterface/interface/prefixScan.h"
#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,276 @@
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h"
+
+namespace cms {
+  namespace alpakatools {
```
```suggestion
namespace cms::alpakatools {
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,276 @@
+
+      using index_type = I;
+
+      struct View {
```
I understand this is a 1:1 translation of the CUDA code, so this clean up couldmalso come later.
Anyway
  - I would prefer to rename this from `View` to `Args` or something that is less confusing with the "View" we have in the SoA and the "View" we have in alpaka :-)
  - in fact, it might be less confusing to just pass the 5 parameters independently, instead of wrapping them in a `struct` ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,276 @@
+          ALPAKA_ASSERT_OFFLOAD(nOnes > 0);
+          int32_t *ppsws = (int32_t *)((char *)(h) + offsetof(OneToManyAssocRandomAccess, psws));
+          auto nthreads = 1024;
+          auto nblocks = (nOnes + nthreads - 1) / nthreads;
```
```suggestion
          auto nblocks = divide_up_by(nOnes, nthreads);
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h:
```diff
@@ -0,0 +1,276 @@
+  }  // namespace alpakatools
+}  // namespace cms
+
+#endif  // HeterogeneousCore_CUDAUtilities_interface_HistoContainer_h
```
```suggestion
#endif  // HeterogeneousCore_AlpakaInterface_interface_OneToManyAssoc_h
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc:
```diff
@@ -0,0 +1,212 @@
+  auto const& devices = cms::alpakatools::devices<Platform>();
+  // auto const& host = cms::alpakatools::host();
+
+  if (devices.empty()) {
+    std::cout << "No devices available on the platform " << EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE)
+              << ", the test will be skipped.\n";
+    return 0;
+  }
+
```
Should this be removed, and scram be made to run the test only if a device is available ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc:
```diff
@@ -0,0 +1,212 @@
+
+    // Running kernel with 1 block, and bs threads per block or elements per thread.
+    // NB: obviously for tests only, for perf would need to use bs = 1024 in GPU version.
+    for (int bs = 32; bs <= 1024; bs += 32) {
```
Would it make sense to use `warpSize` instead of `32` ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc:
```diff
@@ -0,0 +1,212 @@
+      const auto nBlocksInit = divide_up_by(num_items, nThreadsInit);
+      const auto workDivMultiBlockInit = make_workdiv<Acc1D>(nBlocksInit, nThreadsInit);
+
+      alpaka::enqueue(queue,
```
```c++
alpaka::exec(...
```

fwyzard replied:
also below

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc:
Everywhere:
  - replace alpaka::enqueue(queue, alpaka::createTaskKernel<...>(...` with `alpaka::exec<Acc1D>(queue, alpaka::createTaskKernel, ...`
  - I prefer `for (uint32_t i: elements_with_stride(acc, nt)) { ... }` to `  for_each_element_in_grid_strided(acc, n, [&](uint32_t i) { ... });`

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc:
```diff
@@ -0,0 +1,324 @@
+int main() {
+  // get the list of devices on the current platform
+  auto const& devices = cms::alpakatools::devices<Platform>();
+  if (devices.empty()) {
+    std::cout << "No devices available on the platform " << EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE)
+              << ", the test will be skipped.\n";
+    return 0;
+  }
```
remove ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc:
```diff
@@ -0,0 +1,178 @@
+
+int main() {
+  auto const& devices = cms::alpakatools::devices<Platform>();
+  if (devices.empty()) {
+    std::cout << "No devices available on the platform " << EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE)
+              << ", the test will be skipped.\n";
+    return 0;
+  }
+
```
remove ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h:
```diff
@@ -0,0 +1,142 @@
+    // thread-safe version of the vector, when used in a CUDA kernel
+    template <typename TAcc>
+    ALPAKA_FN_ACC int push_back(const TAcc &acc, const T &element) {
+      auto previousSize = alpaka::atomicAdd(acc, &m_size, 1, alpaka::hierarchy::Blocks{});
```
do we have `alpaka::atomicInc(...)` instead of `alpaka::atomicAdd(..., 1)` ?

fwyzard replied:
I assume you meant

> will _not_ make the code any simpler.

While the code will not be simpler, at least on CUDA the performance of `atomicInc` is potentially better than that of `atomicAdd`, because the operations on the same variable in the same warp are implemented using warp primitives, avoiding the contention.

fwyzard replied:
Actually, I have to correct myself: looking at the output of compiler explorer for `atomicInc(ptr, 0xffffffff)` and `atomicAdd(ptr, 1)`, they generate exactly the same SASS code, both for global and shared memory.

So feel free to use the one that you prefer.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/VecArray.h:
```diff
@@ -0,0 +1,107 @@
+      }
+    }
+
+    inline constexpr T &back() const {
```
```suggestion
    inline constexpr T const& back() const {
```
?

ericcano replied:
Indeed. (I will add the mutable version too)

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/VecArray.h:
```diff
@@ -0,0 +1,107 @@
+        return T();  //undefined behaviour
+    }
+
+    // thread-safe version of the vector, when used in a CUDA kernel
```
```suggestion
    // thread-safe version of the vector, when used in a kernel
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc:
```diff
@@ -0,0 +1,255 @@
+      ALPAKA_ASSERT_OFFLOAD(!l);
+    }
+  }
+  int status;
+  auto* demangled = abi::__cxa_demangle(typeid(Hist).name(), NULL, NULL, &status);
+  status || printf("Check contents OK with %s\n", demangled);
+  std::free(demangled);
```
```suggestion
  std::cout << "Check contents OK with " << alpaka::core::demangled<Hist> << '\n';
```

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc:
```diff
@@ -0,0 +1,178 @@
+    ALPAKA_ASSERT_OFFLOAD(N == hist.size());
+
+    // bin
+#ifndef NDEBUG
```
Do we break anything by removing all the `#ifndef NDEBUG` ?
I don't think we ever define `NDEBUG` in CMSSW anyway.

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/radixSort.h:
```diff
@@ -0,0 +1,250 @@
+            typename RF>
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE void radixSortImpl(
+      const TAcc& acc, T const* __restrict__ a, uint16_t* ind, uint16_t* ind2, uint32_t size, RF reorder) {
+#if (defined(ALPAKA_ACC_GPU_CUDA_ENABLED) && defined(__CUDA_ARCH__)) || \
+    (defined(ALPAKA_ACC_GPU_HIP_ENABLED) && defined(__HIP_DEVICE_COMPILE__))
```
what happens if this `#if` is not true ?
that is, what happens on CPU ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/radixSort.h:
```diff
@@ -0,0 +1,250 @@
+        auto laneId = idx & 0x1f;
+
+        for (int offset = 1; offset < 32; offset <<= 1) {
+#if defined(__CUDA_ARCH__)
```
can you add a `// TODO` here, to use the alpaka equivalent as soon as it's available ?
(see https://github.com/alpaka-group/alpaka/pull/1924)

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/radixSort.h:
```diff
@@ -0,0 +1,250 @@
+        auto x = c[idx];
+        auto laneId = idx & 0x1f;
+
+        for (int offset = 1; offset < 32; offset <<= 1) {
```
`32` should be replaced by the actual warp size ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/radixSort.h:
```diff
@@ -0,0 +1,250 @@
+      // prefix scan "optimized"???...
+      for_each_element_in_block(acc, sb, [&](uint32_t idx) {
+        auto x = c[idx];
+        auto laneId = idx & 0x1f;
```
0x1f should be `warp size - 1` ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/radixSort.h:
```diff
@@ -0,0 +1,250 @@
+      alpaka::syncBlockThreads(acc);
+
+      for_each_element_in_block(acc, sb, [&](uint32_t idx) {
+        auto ss = (idx / 32) * 32 - 1;
```
32 vs warp size ?

fwyzard commented on HeterogeneousCore/AlpakaInterface/interface/workdivision.h:
do you have any differences in this file with respect to the version from Adriano ?

smuzaffar commented:
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

AdrianoDee commented:
I was doing some profiling for the pixel track reco with this included and it's pretty slower than before. Doing some profiling I see that the API calls are dominated by `cudaGetDeviceProperties_v2_v12000` that adds up to the "not negligible" ~134s over 2000 events. 

```
 ** CUDA API Summary (cuda_api_sum):

 Time (%)  Total Time (ns)  Num Calls   Avg (ns)     Med (ns)    Min (ns)   Max (ns)   StdDev (ns)                Name               
 --------  ---------------  ---------  -----------  -----------  ---------  ---------  -----------  ---------------------------------
     93.5     134353514397       6002   22384790.8   22648654.5    1357399  218087112   10867341.8  cudaGetDeviceProperties_v2_v12000
      1.9       2679179584      96001      27907.8       6856.0       3608   57418849     301081.7  cudaLaunchKernel                 
      0.9       1354976834       4088     331452.3     338209.5       4277   18971985     610225.3  cudaFreeHost                     
      0.7        987533270       4088     241568.8      13840.0       2983  195419764    3307287.2  cudaHostAlloc                    
      0.7        938885878      74003      12687.1       1207.0        467    9590518     171327.4  cudaEventRecord                  
      0.5        767427057        516    1487261.7      33449.5        544   49158076    4984405.7  cudaEventCreateWithFlags         
      0.5        689040013        346    1991445.1     149592.0       3036   29436752    3804159.3  cudaMalloc                       
      0.4        644311871      26003      24778.4       9138.0       3252    9808336     207163.2  cudaMemcpyAsync                  
      0.3        440782749      10000      44078.3      10150.5       4175   56742897     673947.5  cudaStreamAddCallback            
      0.2        248554459      89098       2789.7       2002.0        784    1185539       5423.0  cudaEventQuery                   
```

Also in the visual profiler:

<img width="1470" alt="image" src="https://github.com/cms-sw/cmssw/assets/16901146/43547a9f-9147-4220-90a6-5af40a3eaf58">


Doing some investigations in the Alpaka codebase (I'm not sure I'm pointing to the right one but I imagine it still applies) it appears that this is used in [`getDeviceProperties`](https://github.com/alpaka-group/alpaka/blob/ac4e16feefbfd1976e3c908dcc8560accbeed6f6/include/alpaka/alpaka.hpp#L24909-L24913) that is used in [`getWarpSizes`](https://github.com/alpaka-group/alpaka/blob/ac4e16feefbfd1976e3c908dcc8560accbeed6f6/include/alpaka/alpaka.hpp#L23793-L23803).

And then this is used in `OneToManyAssocRandomAccess:: launchFinalize`

https://github.com/cms-sw/cmssw/blob/5f9c2e62c82c8f5a6ea6cb93a2e67bb3ec564388/HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h#L266

The smoking gun comes from the profiling of `alpakaTestHistoContainerCudaAsync` that reads:

```
 ** CUDA API Summary (cuda_api_sum):

 Time (%)  Total Time (ns)  Num Calls  Avg (ns)   Med (ns)   Min (ns)  Max (ns)  StdDev (ns)                Name               
 --------  ---------------  ---------  ---------  ---------  --------  --------  -----------  ---------------------------------
     84.7         35343987         21  1683047.0  1683270.0   1643330   1794078      36393.7  cudaGetDeviceProperties_v2_v12000
      8.4          3495506         80    43693.8     5930.5      4709   2781034     310026.2  cudaLaunchKernel                 
      2.8          1184093          6   197348.8    12969.5      2603   1123658     453890.2  cudaHostAlloc                    
      1.2           484352          6    80725.3     4320.0      2936    461249     186424.4  cudaFreeHost                     
      0.8           350542         52     6741.2     5447.5      2889     31396       5220.6  cudaMemcpyAsync                  
      0.5           227710         20    11385.5     8027.0      6108     19323       5967.4  cudaStreamSynchronize            
```

With the poor man solution for CUDA (I don't if there's already a way to get the proper number of threads per warp  for other backends):

```diff
diff --git a/HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h b/HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h
index d4e3a65ae5d..c1ea4fed8e3 100644
--- a/HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h
+++ b/HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h
@@ -251,7 +251,14 @@ namespace cms {
           int32_t *ppsws = (int32_t *)((char *)(h) + offsetof(OneToManyAssocRandomAccess, psws));
           auto nthreads = 1024;
           auto nblocks = (nOnes + nthreads - 1) / nthreads;
-          auto workDiv = cms::alpakatools::make_workdiv<TAcc>(nblocks, nthreads);
+          auto workDiv = cms::alpakatools::make_workdiv<TAcc>(nblocks, nthreads); 
+          
+          auto numberOfWarps = 32; 
+          #ifndef ALPAKA_ACC_GPU_CUDA_ENABLED
+                 numberOfWarps = alpaka::getWarpSizes(alpaka::getDev(queue))[0]; 
+          #endif
+
           alpaka::exec<TAcc>(queue,
                              workDiv,
                              multiBlockPrefixScan<Counter>(),
@@ -260,7 +267,7 @@ namespace cms {
                              nOnes,
                              nblocks,
                              ppsws,
-                             alpaka::getWarpSizes(alpaka::getDev(queue))[0]);
+                             numberOfWarps);
         } else {
           h->finalize();
         }
```

the profiling for `alpakaTestHistoContainerCudaAsync` is back to normal:

```
 ** CUDA API Summary (cuda_api_sum):

 Time (%)  Total Time (ns)  Num Calls  Avg (ns)   Med (ns)   Min (ns)  Max (ns)  StdDev (ns)                Name               
 --------  ---------------  ---------  ---------  ---------  --------  --------  -----------  ---------------------------------
     43.9          3360218         80    42002.7     3818.5      3179   2879851     321356.7  cudaLaunchKernel                 
     19.1          1466118          1  1466118.0  1466118.0   1466118   1466118          0.0  cudaGetDeviceProperties_v2_v12000
     16.1          1231870          6   205311.7     8570.0      2514   1195633     485166.8  cudaHostAlloc                    
      6.5           497739          6    82956.5     3956.5      2117    477765     193420.5  cudaFreeHost                     
      3.6           279500         52     5375.0     4315.5      2501     38103       5110.4  cudaMemcpyAsync                  
      2.7           204737         20    10236.9     7120.0      4530     22421       6025.4  cudaStreamSynchronize            
      2.4           180809         30     6027.0     4231.5      3319     29519       4940.4  cudaMemsetAsync                  
      2.2           167588          6    27931.3     3333.5      2265    147525      58629.2  cudaFree                         
```

VinInn commented:
This is one of the reason I got rid of CUB and implemented prefixscan and sort by myself...

makortel commented:
Good catch @AdrianoDee. Given that Alpaka Devices are already objects, I'd much prefer they'd do these kind of queries once and cache the result (either in the Device constructor, or lazily on first request). (or we come up with our own caching mechanism)

In the mean time, perhaps a better workaround would be an abstracted function (in `HeterogeneousCore/AlpakaInterface`) that returns 1 / 32 / (/ 64?) / `alpaka::getWarpSizes()` for CPU / NVIDIA (/ AMD?) / default?

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37871

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37871/cms-checkout-topic.log

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37873

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37873/cms-checkout-topic.log

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37874

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37874/cms-checkout-topic.log

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/37875

- This PR adds an extra 72KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117, #43164
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117

cmsbuild commented:
Pull request #43064 was updated. @cmsbuild, @fwyzard, @makortel can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/38067

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117

cmsbuild commented:
Pull request #43064 was updated. @fwyzard, @makortel, @cmsbuild can you please check and sign again.

ericcano commented:
Investigating the failure of the serial version of the prefix scan unit test.

cmsbuild commented:
Pull request #43064 was updated. @makortel, @fwyzard, @cmsbuild can you please check and sign again.

fwyzard committed:
Moves from elements_with_stride to independent_group_elements

ericcano committed:
Final fixes to unit tests of port to Alpaka

Reinstates error messages instead of silent call to `REQUIRE(!devices.empty());`
Adds maybe_unused attirbut for variable used in assert.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/38475

- This PR adds an extra 60KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117

cmsbuild commented:
Pull request #43064 was updated. @cmsbuild, @fwyzard, @makortel can you please check and sign again.

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testIndependentKernel.dev.cc:
```diff
@@ -121,10 +121,7 @@ TEST_CASE("Test alpaka kernels for the " EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESP
-      INFO("No devices available on the platform " EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE));
-      REQUIRE(not devices.empty());
-    }
+    REQUIRE(!devices.empty());
```
Can you revert this change ?
I liked the more explicit explanatory message better.

fwyzard commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testKernel.dev.cc:
```diff
@@ -377,10 +377,7 @@ TEST_CASE("Test alpaka kernels for the " EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESP
-      INFO("No devices available on the platform " EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE));
-      REQUIRE(not devices.empty());
-    }
+    REQUIRE(!devices.empty());
```
Can you revert this change ?
I liked the more explicit explanatory message better.

ericcano replied:
Sure, I reinstated them.

fwyzard commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   HeaderConsistency  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0bcd78/36941/summary.html
**COMMIT**: 40911b448534b768669fea307a8dc8a5199414a1
**CMSSW**: CMSSW_14_0_X_2024-01-19-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43064/36941/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/38488

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/38490

- This PR adds an extra 72KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117

cmsbuild commented:
Pull request #43064 was updated. @fwyzard, @makortel, @cmsbuild can you please check and sign again.

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0bcd78/36964/summary.html
**COMMIT**: 70c3404d5f575297019a20aa7bc162fdf1d9e2a3
**CMSSW**: CMSSW_14_0_X_2024-01-22-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43064/36964/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/38497

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117

fwyzard commented:
@ericcano let's see if the bot is finally happy :-)

Can you squash the last three commits into 40911b44853 ?

cmsbuild commented:
Pull request #43064 was updated. @makortel, @fwyzard can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43064/38498

- This PR adds an extra 64KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/FlexiStorage.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/OneToManyAssoc.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/VecArray.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/alpakastdAlgorithm.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/prefixScan.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/radixSort.h  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/interface/workdivision.h  modified in PR(s): #40932, #41117
  - File  HeterogeneousCore/AlpakaInterface/test/BuildFile.xml  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneHistoContainer.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testOneToManyAssoc.dev.cc  modified in PR(s): #41117
  - File  HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc  modified in PR(s): #41117

cmsbuild commented:
Pull request #43064 was updated. @makortel, @cmsbuild, @fwyzard can you please check and sign again.

fwyzard commented:
enable gpu

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @antoniovilela, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
